From: Krzysztof Mazur <krzysiek@podlesie.net>
Date: Wed, 19 Dec 2012 11:03:41 +0100
Subject: i915: ensure that VGA plane is disabled

commit 0fde901f1ddd2ce0e380a6444f1fb7ca555859e9 upstream.

Some broken systems (like HP nc6120) in some cases, usually after LID
close/open, enable VGA plane, making display unusable (black screen on LVDS,
some strange mode on VGA output). We used to disable VGA plane only once at
startup. Now we also check, if VGA plane is still disabled while changing
mode, and fix that if something changed it.

Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=57434
Signed-off-by: Krzysztof Mazur <krzysiek@podlesie.net>
Signed-off-by: Daniel Vetter <daniel.vetter@ffwll.ch>
[bwh: Backported to 3.2: intel_modeset_setup_hw_state() does not
 exist, so call i915_redisable_vga() directly from intel_lid_notify()]
Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
---
Index: linux-3.2.46/drivers/gpu/drm/i915/intel_display.c
===================================================================
--- linux-3.2.46.orig/drivers/gpu/drm/i915/intel_display.c	2013-07-26 19:27:26.000000000 +0000
+++ linux-3.2.46/drivers/gpu/drm/i915/intel_display.c	2013-07-26 19:27:30.000000000 +0000
@@ -9314,6 +9314,23 @@
 	POSTING_READ(vga_reg);
 }
 
+void i915_redisable_vga(struct drm_device *dev)
+{
+	struct drm_i915_private *dev_priv = dev->dev_private;
+	u32 vga_reg;
+
+	if (HAS_PCH_SPLIT(dev))
+		vga_reg = CPU_VGACNTRL;
+	else
+		vga_reg = VGACNTRL;
+
+	if (I915_READ(vga_reg) != VGA_DISP_DISABLE) {
+		DRM_DEBUG_KMS("Something enabled VGA plane, disabling it\n");
+		I915_WRITE(vga_reg, VGA_DISP_DISABLE);
+		POSTING_READ(vga_reg);
+	}
+}
+
 void intel_modeset_init(struct drm_device *dev)
 {
 	struct drm_i915_private *dev_priv = dev->dev_private;
Index: linux-3.2.46/drivers/gpu/drm/i915/i915_drv.h
===================================================================
--- linux-3.2.46.orig/drivers/gpu/drm/i915/i915_drv.h	2013-07-26 19:27:17.000000000 +0000
+++ linux-3.2.46/drivers/gpu/drm/i915/i915_drv.h	2013-07-26 19:27:30.000000000 +0000
@@ -1397,6 +1397,7 @@
 #endif /* CONFIG_ACPI */
 
 /* modesetting */
+extern void i915_redisable_vga(struct drm_device *dev);
 extern void intel_modeset_init(struct drm_device *dev);
 extern void intel_modeset_gem_init(struct drm_device *dev);
 extern void intel_modeset_cleanup(struct drm_device *dev);
Index: linux-3.2.46/drivers/gpu/drm/i915/intel_lvds.c
===================================================================
--- linux-3.2.46.orig/drivers/gpu/drm/i915/intel_lvds.c	2013-07-26 19:26:22.000000000 +0000
+++ linux-3.2.46/drivers/gpu/drm/i915/intel_lvds.c	2013-07-26 19:27:30.000000000 +0000
@@ -541,6 +541,7 @@
 
 	mutex_lock(&dev->mode_config.mutex);
 	drm_helper_resume_force_mode(dev);
+	i915_redisable_vga(dev);
 	mutex_unlock(&dev->mode_config.mutex);
 
 	return NOTIFY_OK;
